### PR TITLE
LPS-199231 Added new autogenerated method that returns an atribute by it's name + usage for declaring the model attributes in API Builder

### PR DESCRIPTION
**Pull request by jalber786**
**What's new?:**
- Created new autogenerated method called "getValue" in the DTOs which returns a specific atribute by it's name. It takes into account both model attributes and the "properties" defined by a json-map.
- Executed buildREST in all the portal.

**What's changed?:**
- Swapped the previous reflection usage with the new "getValue" method.

See the following **Jira ticket**: [LPS-198749](https://liferay.atlassian.net/browse/LPS-198749)

### CODE REVIEW

**Comment by jalber786** [comment]
ci:test:relevant

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl [hunk 1]
```diff
@@ -139,6 +141,10 @@ public class ${schemaName} <#if dtoParentClassName?has_content>extends ${dtoPare
 			sizeParameters = []
 		/>
 
+		<#if !stringUtil.equals(propertyName, "properties")>
```
Comment (ccorreagg) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl:144]: What is the purpose of this empty if clause? :thinking:

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl [hunk 2]
```diff
@@ -129,6 +129,8 @@ public class ${schemaName} <#if dtoParentClassName?has_content>extends ${dtoPare
 
 	<#assign
 		enumSchemas = freeMarkerTool.getDTOEnumSchemas(openAPIYAML, schema)
+		generateGetValues = []
```
Comment (ccorreagg) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl:132]: This name is confusing. To me it sounds as it should be a boolean

File: modules/apps/analytics/analytics-dxp-entity-rest-api/src/main/java/com/liferay/analytics/dxp/entity/rest/dto/v1_0/DXPEntity.java [hunk 3]
```diff
@@ -206,6 +206,30 @@ public boolean equals(Object object) {
 		return Objects.equals(toString(), dxpEntity.toString());
 	}
 
+	public Object getValue(String propertyName) {
```
Comment (ccorreagg) [inline modules/apps/analytics/analytics-dxp-entity-rest-api/src/main/java/com/liferay/analytics/dxp/entity/rest/dto/v1_0/DXPEntity.java:209]: I think It does not make sense to generate this method for the DTOs that do not have any x-json-map property, at least by now.

For our use case, this method has to be generated only for the cases where x-json-map is true

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl [hunk 4]
```diff
@@ -281,6 +290,27 @@ public class ${schemaName} <#if dtoParentClassName?has_content>extends ${dtoPare
 		return Objects.equals(toString(), ${schemaVarName}.toString());
 	}
 
+	<#if generateGetValues?has_content>
+		<#assign firstPropertyName = generateGetValues[0]>
+			public Object getValue(String propertyName) {
+				if (Objects.equals(propertyName, "${firstPropertyName}")) {
+					return ${firstPropertyName};
+		<#assign reducedProperties = generateGetValues[0..generateGetValues?size - 1]>
+		<#list reducedProperties as propertyName>
```
Comment (ccorreagg) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl:299]: This is very complex. I think we should be able to include all the cases in a for loop for better understanding

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl [hunk 5]
```diff
@@ -281,6 +290,27 @@ public class ${schemaName} <#if dtoParentClassName?has_content>extends ${dtoPare
 		return Objects.equals(toString(), ${schemaVarName}.toString());
 	}
 
+	<#if generateGetValues?has_content>
```
Comment (ccorreagg) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl:293]: I think we need to update this condition, as we don't want to generate for all the entities, only for the ones that has a property with `x-json-map` set

**Review by ccorreagg — CHANGES_REQUESTED** [review]
Please have a look at my comments

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl [hunk 6]
```diff
@@ -281,6 +290,27 @@ public class ${schemaName} <#if dtoParentClassName?has_content>extends ${dtoPare
...
+		<#assign firstPropertyName = generateGetValues[0]>
+			public Object getValue(String propertyName) {
```
Comment (ccorreagg) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl:295]: I think the commit and the PR description is wrong. The method is called getValue

**Comment by ccorreagg** [comment]
@jalber786 in the latest commit, we could update the commit message from "LPS-199231 Introduced the usage of the getValue method" to "LPS-199231 Use new method and avoid using reflection". I think it would explain better the purpose of the PR and maybe Brian will accept the PR in a better way

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl (hunk 1 above)
Comment (jalber786) [reply modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl:144]: It was a old if statement that I used before some changes, I forgot deleting it sorry.

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl [hunk 7]
```diff
@@ -281,6 +290,27 @@ public class ${schemaName} <#if dtoParentClassName?has_content>extends ${dtoPare
...
+				if (Objects.equals(propertyName, "${firstPropertyName}")) {
+					return ${firstPropertyName};
+		<#assign reducedProperties = generateGetValues[0..generateGetValues?size - 1]>
+		<#list reducedProperties as propertyName>
+				} else if (Objects.equals(propertyName, "${propertyName}")) {
+					return ${propertyName};
+		</#list>
+		<#if hasJsonMap>
+ 				} else {
+					return properties.get(propertyName);
```
Comment (ccorreagg) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl:305]: What if the property with the `x-json-map` is not called `properties`?

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl (hunk 4 above)
Comment (ccorreagg) [reply modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/dto.ftl:299]: I think you can use `<#sep> [....] </#sep>` for this purpose.

**Comment by jalber786** [comment]
ci:stop

**Comment by jalber786** [comment]
ci:test:sf

**Comment by jalber786** [comment]
ci:test:relevant

**Review by ccorreagg — COMMENTED** [review]
Could you @jalber786 add a test for this? Have a look at the `RESTBuilderTest` class

**Comment by jalber786** [comment]
ci:test:sf

**Comment by jalber786** [comment]
ci:test:relevant

**Comment by jalber786** [comment]
ci:reevaluate:1357401_4715

**Comment by jalber786** [comment]
ci:test:sf

**Comment by jalber786** [comment]
ci:test:relevant

File: modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml [hunk 8]
```diff
@@ -76,6 +76,26 @@ components:
...
             type: object
+        TestWithJsonMapAttribute:
+            description:
+                Test Component to test the generation of getValue method on DTOs when a Json Map attribute is present.
+            properties:
+                description:
+                    type: string
+                name:
+                    type: string
+                properties:
```
Comment (sergiojimcos) [inline modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml:87]: It's weird one calling `properties` and the other `properties2`. If you are testing both properties, you must follow the nomenclature: `property1` and `porperty2`.

File: modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml [hunk 9]
```diff
@@ -76,6 +76,26 @@ components:
...
+                    additionalProperties:
+                        type: object
+                    type: object
+                    x-json-map: true
+                properties2:
```
Comment (sergiojimcos) [inline modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml:92]: Is it enough to test a second property with json-map?

**Review by sergiojimcos — CHANGES_REQUESTED** [review]
Take a look at my comments/questions

File: modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml (hunk 8 above)
Comment (ccorreagg) [reply modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml:87]: I would say `properties1` and `properties2`

File: modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml [hunk 10]
```diff
@@ -76,6 +76,26 @@ components:
                     type: array
             title: TestArray
             type: object
+        TestWithJsonMapAttribute:
```
Comment (ccorreagg) [inline modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml:79]: Maybe better using a name that matches the previous format, that means without the "With" word: `TestJsonMapAttribute` (it needs to be sorted then)

File: modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml (hunk 8 above)
Comment (jalber786) [reply modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml:87]: Ok! I'm on it!

File: modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml (hunk 10 above)
Comment (jalber786) [reply modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml:79]: Yeah sure thing. I will change it right now.

**Comment by jalber786** [comment]
ci:test:sf

**Comment by jalber786** [comment]
ci:test:relevant

**Review by sergiojimcos — APPROVED** [review]

**Review by ccorreagg — COMMENTED** [review]
I see an error in the last execution of SF:

```
     [java] 2: /opt/dev/projects/github/liferay-portal/modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml expected:<...ltiple JSON Maps are[
     [java]                ] present.
     [java]           ...> but was:<...ltiple JSON Maps are[] present.
     [java]           ...>
```

I think it needs to be fixed before forwarding to BC

**Comment by jalber786** [comment]
> I see an error in the last execution of SF:
> 
> ```
>      [java] 2: /opt/dev/projects/github/liferay-portal/modules/util/portal-tools-rest-builder/src/test/resources/com/liferay/portal/tools/rest/builder/dependencies/rest-openapi.yaml expected:<...ltiple JSON Maps are[
>      [java]                ] present.
>      [java]           ...> but was:<...ltiple JSON Maps are[] present.
>      [java]           ...>
> ```
> 
> I think it needs to be fixed before forwarding to BC

I didn't see it sorry! I made the change so it should be fixed now

**Comment by jalber786** [comment]
ci:test:sf

**Comment by jalber786** [comment]
ci:test:relevant

**Review by ccorreagg — APPROVED** [review]
LGTM!

**Comment by yanan** [comment]
Test failures are not related, ok to forward :partying_face:

**Comment by jalber786** [comment]
ci:forward

**Comment by jalber786** [comment]
ci:forward:force

**Pull request closed without merging**